integration_tests/binding/binding.go

```diff
@@ -0,0 +1,231 @@
+}
+
+var (
+	keep = flag.Bool("keep", false, "Keep topology deployed after test")
+)
+
+// Get returns the custom lemming binding. The topoDir is the relative path to a
+// directory containing the Ondatra testbed and KNE topology pb.txt files.
+func Get(topoDir string) func() (binding.Binding, error) {
+	dir, _ := filepath.Abs(topoDir)
+	testbedFile := filepath.Join(dir, "testbed.pb.txt")
+	topoFile := filepath.Join(dir, "topology.pb.txt")
+
+	flag.Set("testbed", testbedFile)
+
+	return func() (binding.Binding, error) {
+		fmt.Println("Checking for KNE CLI installation")
+		knePath, err := getKNECLIPath()
+		if err != nil {
+			return nil, err
+		}
+
+		fmt.Println("Creating knebind config file")
+		if err := writeKNEBindCfg(knePath, topoFile); err != nil {
+			return nil, err
+		}
+		b, err := kinit.Init()
+		if err != nil {
+			return nil, err
+		}
+		return &LemmingBind{
+			Binding:  b,
+			topoFile: topoFile,
+			kubecfg:  os.ExpandEnv("$HOME/.kube/config"),
+		}, nil
+	}
+}
+
+// LemmingBind wraps the Ondatra knebind and adds the ability to set the testbed and topology from inside the test.
+// TODO: Add unit tests and upstream some of this to Ondatra.
+type LemmingBind struct {
+	binding.Binding
+	topoFile string
+	kubecfg  string
+	created  bool
+	id       string
+	cancel   func()
+}
+
+// Release runs knebind release then deletes the topology if it was created by this binding.
+func (lb *LemmingBind) Release(ctx context.Context) error {
+	defer lb.cancel()
+	if err := lb.Binding.Release(ctx); err != nil {
+		return err
+	}
+	if !lb.created || *keep {
+		return nil
+	}
+	if out, err := exec.Command("kne", "delete", lb.topoFile).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed delete topology: %v output:\n%s", err, string(out))
+	}
+	return nil
+}
+
+// Reserve deploys the topology if it doesn't exists, then runs knebind Reserve.
+func (lb *LemmingBind) Reserve(ctx context.Context, tb *proto.Testbed, runTime time.Duration, waitTime time.Duration, partial map[string]string) (*binding.Reservation, error) {
+	topoBytes, err := os.ReadFile(lb.topoFile)
+	if err != nil {
+		return nil, err
+	}
+	topo := &tpb.Topology{}
+	if err := prototext.Unmarshal(topoBytes, topo); err != nil {
+		return nil, err
+	}
+
+	// Get kubernetes API client.
+	cfg, err := clientcmd.BuildConfigFromFlags("", lb.kubecfg)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Acquire a lock on this topology, prevent another test from using this topology while this test is running.
+	lb.id = uuid.New().String()
+	lock, err := resourcelock.New(
+		resourcelock.ConfigMapsLeasesResourceLock,
+		"default",
+		topo.GetName(),
+		client.CoreV1(),
+		client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{Identity: lb.id},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ready := make(chan bool)
+	elect, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
```

Comment:
The lock will be grabbed until cancel is called? What determines when the lease is released?

Reply:
The lock is held for the duration of the test. The goal is to prevent two tests from running on the same topology at the same time. The lock is release when Ondatra calls binding.Release (after all tests are run).

When you run go test ./... it (by default) runs multiples packages in parallel, so without this tests could run on same duts at the same time.

Reply:
Ok, I assume `LeaseDuration` isn't the upper bound on how long each test can be the leader?

Reply:
correct, the leader refreshes the lock every 10 sec, the non-leaders wait for 15 secs before becoming the leader